include/nbl/ext/ImGui/ImGui.h

```diff
@@ -1,76 +1,82 @@
-		void Text(char const* label, ...);
-		void InputFloat(char const* label, float* value);
-		void InputFloat2(char const* label, float* value);
-		void InputFloat3(char const* label, float* value);
-		void InputFloat4(char const* label, float* value);
-		void InputFloat3(char const* label, nbl::core::vector3df& value);
-		bool Combo(char const* label, int32_t* selectedItemIndex, char const** items, int32_t itemsCount);
-		bool Combo(const char* label, int* selectedItemIndex, std::vector<std::string>& values);
-		void SliderInt(char const* label, int* value, int minValue, int maxValue);
-		void SliderFloat(char const* label, float* value, float minValue, float maxValue);
-		void Checkbox(char const* label, bool* value);
-		void Spacing();
-		void Button(char const* label, std::function<void()> const& onPress);
-		void InputText(char const* label, std::string& outValue);
-		[[nodiscard]] bool HasFocus();		
-		[[nodiscard]] bool IsItemActive();
-		[[nodiscard]] bool TreeNode(char const* name);
-		void TreePop();
+		//! Nabla ImGUI backend reserves this index for font atlas, any attempt to hook user defined texture within the index will cause runtime error [TODO: could have a setter & getter to control the default & currently hooked font texture ID and init 0u by default]
+		_NBL_STATIC_INLINE_CONSTEXPR auto NBL_FONT_ATLAS_TEX_ID = 0u;
 
-	private:
+		//! update ImGUI internal state & cpu draw command lists, call it before this->render
+		bool update(const ui::IWindow* window, float deltaTimeInSec, const core::SRange<const nbl::ui::SMouseEvent> mouseEvents, const core::SRange<const nbl::ui::SKeyboardEvent> keyboardEvents);
+
+		//! updates mapped mdi buffer & records draw calls [TODO: maybe its a good idea to update mdi at the end of .update call instead of doing it here?]
+		bool render(nbl::video::SIntendedSubmitInfo& info, const nbl::video::IGPUDescriptorSet* const descriptorSet);
+
+		//! registers lambda listener in which ImGUI calls should be recorded
+		int registerListener(std::function<void()> const& listener);
+		bool unregisterListener(uint32_t id);
 
-		core::smart_refctd_ptr<video::IGPUDescriptorSetLayout> CreateDescriptorSetLayout();
+		//! sets ImGUI context, you are supposed to pass valid ImGuiContext* context
+		void setContext(void* imguiContext);
 
-		void CreatePipeline(video::IGPURenderpass* renderpass, video::IGPUPipelineCache* pipelineCache);
-		// TODO: just take an intended next submit instead of queue and cmdbuf, so we're consistent across utilities
-		video::ISemaphore::future_t<video::IQueue::RESULT> CreateFontTexture(video::IGPUCommandBuffer* cmdBuffer, video::IQueue* queue);
-		void UpdateDescriptorSets();
+		//! image view getter to access default font texture
+		inline auto getFontAtlasView() -> decltype(auto) { return core::smart_refctd_ptr(m_fontAtlasTexture); }
+
+		//! mdi streaming buffer getter
+		inline auto getStreamingBuffer() -> decltype(auto) { return (std::as_const(m_mdi.streamingTDBufferST)); }
```

Comment:
getters should always return dumb pointers, never smart because just a query shouldn't automatically grant ownership